src/wav2vec2/modeling.py

```diff
@@ -0,0 +1,311 @@
+# __author__ = "Vasudev Gupta"
+# __author_email__ = "7vasudevgupta@gmail.com"
+
+"""TensorFlow implementation of Wav2Vec2"""
+
+import os
+import subprocess
+from dataclasses import replace
+
+import numpy as np
+import tensorflow as tf
+from huggingface_hub import ModelHubMixin
+
+from .config import Wav2Vec2Config
+from .tensorflow_addons import Conv1DWithWeightNorm, GroupNormalization
+
+
+class TransformerAttention(tf.keras.layers.Layer):
+    """Attention layer from `Attention Is All You Need`"""
+
+    def __init__(self, config, name="attention"):
+        super().__init__(name=name)
+        self.num_heads = config.num_heads
+
+        self.q = tf.keras.layers.Dense(config.hidden_size, name="q_proj")
+        self.k = tf.keras.layers.Dense(config.hidden_size, name="k_proj")
+        self.v = tf.keras.layers.Dense(config.hidden_size, name="v_proj")
+        self.projection = tf.keras.layers.Dense(config.hidden_size, name="out_proj")
+
+        self.dropout = tf.keras.layers.Dropout(config.dropout)
+
+    def call(self, batch, training=False):
+        head_size = batch.shape[2] // self.num_heads
+        q_out = self._prepare_either_qkv(self.q(batch), head_size)
+        k_out = self._prepare_either_qkv(self.k(batch), head_size)
+        v_out = self._prepare_either_qkv(self.v(batch), head_size)
+
+        q_out = q_out * head_size ** (-0.5)
+
+        batch = self.get_context(q_out, k_out, v_out, training=training)
+        batch = self.projection(batch)
+        return batch
+
+    def get_context(self, q_out, k_out, v_out, training=False):
+
+        b, h, l, d = q_out.shape
+        attn_scores = tf.matmul(q_out, k_out, transpose_b=True)  # "bhqd,bhkd->bhqk"
+
+        attn_scores = self.dropout(
+            tf.nn.softmax(attn_scores, axis=-1), training=training
+        )
+        context = tf.matmul(attn_scores, v_out)  # "bhll,bhld->bhld"
+        context = tf.transpose(context, perm=(0, 2, 1, 3))
+        return tf.reshape(context, (-1, l, h * d))
+
+    def _prepare_either_qkv(self, tensor, head_size):
+        bsz, seqlen, _ = tensor.shape
+        tensor = tf.reshape(tensor, (-1, seqlen, self.num_heads, head_size))
+        return tf.transpose(
+            tensor, perm=(0, 2, 1, 3)
+        )  # -> bsz, num_heads, seqlen, head_size
```

Comment:
Prefer implementing the `get_config()` layer too. This helps in serializing a particular layer of a model and also the entire model and then loading them back. Since this is a custom layer, it's important that we do this. Get an overview [here](https://livebook.manning.com/book/deep-learning-with-python-second-edition/chapter-11/).